### PR TITLE
Remove `disable_relative_sampling` option.

### PR DESCRIPTION
**Pull request by sile**
This PR removes `disable_relative_sampling` option (it was introduced by #380 as a workaround; see also https://github.com/pfnet/optuna/pull/427#issuecomment-501993365).

To make this option unnecessary, this PR contains the following changes:
- Make `ChainerMNTrial` inherit `BaseTrial` instead of `Trial`
- Change the type of the first argument of the objective function for `ChainerMNStudy` from `Trial` to `ChainerMNTrial`

### CODE REVIEW

**Comment by c-bata** [comment]
I agree with removing `disable_relative_sampling` option. Although #380 is merged, is this PR going to be merged?

If we are using a Sampler that implements the `sample_relative` method, we have to be careful that this option is not True. Even though the user is using GPBO sampler, if this option becomes True, all parameters will be sampled by sample_independent.

I may not know exactly what [this commit](https://github.com/pfnet/optuna/pull/380/files/5b9a5aeef9b5b0b5952671403fafb646112b2bf9..7909c237dfbe0217d8b5ee4c20d6eb7ca2f95f7c) is intended, but providing two abstract base classes, such as `BaseSampler` and `BaseRelativeSampler`, and judging with the `isinstance` function may be one solution.

**Comment by sile** [comment]
> Although #380 is merged, is this PR going to be merged?

Yes, we want to merge this PR before the next release.
`disable_relative_sampling` option is a workaround to avoid a problem when using `ChainerMNStudy` with the new sampler interface. In this PR, I will solve the problem itself and make the option unnecessary.

Thank you for your kind advice.

File: optuna/integration/chainermn.py [hunk 1]
```diff
@@ -168,66 +167,67 @@ def __setattr__(self, attr_name, value):
...
-        return self._suggest_with_mpi(self.delegate.suggest_categorical, *(name, choices))
+        def func():
+            # type: () -> T
+
+            assert self.delegate is not None
+            return self.delegate.suggest_categorical(name, choices)
+
+        return self._call_with_mpi(func)
 
     def _suggest_with_mpi(self, func, *args):
```
Comment (toshihikoyanase) [inline optuna/integration/chainermn.py]: `_suggest_with_mpi` is not called from anywhere, and I think we can remove it.

Comment (sile) [reply optuna/integration/chainermn.py]: Oops..., thank you for pointing it out!

Fixed: https://github.com/pfnet/optuna/pull/427/commits/3e4cef8f4a95c1c22cc34548e7be3f62a38dcd8f

File: optuna/integration/chainermn.py [hunk 2]
```diff
@@ -25,6 +21,11 @@
     from typing import TypeVar  # NOQA
     from typing import Union  # NOQA
 
+    from optuna.distributions import BaseDistribution  # NOQA
+    from optuna.storages import BaseStorage  # NOQA
```
Comment (toshihikoyanase) [inline optuna/integration/chainermn.py]: Please remove this line because `BaseStorage` is not used in this file.

Comment (sile) [reply optuna/integration/chainermn.py]: Thanks.
Removed: https://github.com/pfnet/optuna/pull/427/commits/9b1828243a2145cbb419f78d682ae34e2fbc760c

File: optuna/integration/chainermn.py [hunk 3]
```diff
@@ -168,157 +166,190 @@ def __setattr__(self, attr_name, value):
...
-                raise
-        else:
-            result = self.comm.mpi_comm.bcast(None)
-            if isinstance(result, Exception):
-                raise result
-            return result
-
-
-class _ChainerMNTrial(Trial):
+class ChainerMNTrial(BaseTrial):
```
Comment (toshihikoyanase) [inline optuna/integration/chainermn.py]: In this PR, `ChainerMNTrial` will be a public class, and users may want to consult its documentation when they write objective functions.
Could you add documentation for `ChainerMNTrial`?

The following is just an example:
```
    """A wrapper of :class:`~optuna.trial.Trial` to incorporate Optuna with ChainerMN.

    .. seealso::
        :class:`~optuna.integration.chainermn.ChainerMNTrial` provides the same interface as
        :class:`~optuna.trial.Trial`. Please refer to :class:`optuna.trial.Trial` for further
        details.

    Args:
        trial:
            A :class:`~optuna.trial.Trial` object.
        comm:
            A `ChainerMN communicator <https://docs.chainer.org/en/stable/chainermn/reference/
            index.html#communicators>`_.
    """

```

Comment (sile) [reply optuna/integration/chainermn.py]: Added: https://github.com/pfnet/optuna/pull/427/commits/e67caf1fe67ef3982fbfcd47243cbeb5cd28e6b4

Thank you for your suggestion.

File: tests/integration_tests/test_chainermn.py [hunk 4]
```diff
@@ -25,6 +25,7 @@
     from typing import Optional  # NOQA
     from typing import Type  # NOQA
 
+    from optuan.integration.chainermn import ChainerMNTrial  # NOQA
```
Comment (toshihikoyanase) [inline tests/integration_tests/test_chainermn.py]: `optuan.integration.chainermn` seems to be a typo for `optuna.integration.chainermn`.

Comment (sile) [reply tests/integration_tests/test_chainermn.py]: Thanks. Fixed: https://github.com/pfnet/optuna/pull/427/commits/734b17eb89ff1b7b43fd0e6a67e069f71b8eea1a

**Pull request merged**